cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/task/ProvisioningTask.java

```diff
@@ -0,0 +1,140 @@
+   * @throws Exception if there was a non-retryable exception while executing a subtask
+   */
+  public void execute() throws Exception {
+    RetryStrategy retryStrategy =
+      RetryStrategies.statefulTimeLimit(retryTimeLimitSecs, TimeUnit.SECONDS, System.currentTimeMillis(),
+                                        RetryStrategies.exponentialDelay(100, 20000, TimeUnit.MILLISECONDS));
+    Map<ProvisioningOp.Status, ProvisioningSubtask> subtasks = getSubtasks();
+
+    Optional<ProvisioningTaskInfo> taskInfoOptional = Optional.of(initialTaskInfo);
+    while (taskInfoOptional.isPresent()) {
+      ProvisioningTaskInfo taskInfo = taskInfoOptional.get();
+      persistTaskInfo(taskInfo, retryStrategy);
+
+      ProvisioningOp.Status state = taskInfo.getProvisioningOp().getStatus();
+      ProvisioningSubtask subtask = subtasks.get(state);
+      if (subtask == null) {
+        // should never happen
+        throw new IllegalStateException(
+          String.format("Invalid state '%s' in provisioning task for program run '%s'. "
+                          + "This means there is a bug in provisioning state machine. "
+                          + "Please reach out to the development team.",
+                        state, programRunId));
+      }
+
+      try {
+        taskInfoOptional = Retries.callWithInterruptibleRetries(() -> subtask.execute(taskInfo), retryStrategy,
+                                                                t -> t instanceof RetryableProvisionException);
+      } catch (InterruptedException e) {
+        throw e;
+      } catch (Exception e) {
+        handleSubtaskFailure(taskInfo, e);
+        ProvisioningOp failureOp = new ProvisioningOp(taskInfo.getProvisioningOp().getType(),
+                                                      ProvisioningOp.Status.FAILED);
+        ProvisioningTaskInfo failureInfo = new ProvisioningTaskInfo(taskInfo, failureOp, taskInfo.getCluster());
+        persistTaskInfo(failureInfo, retryStrategy);
+        return;
+      }
+    }
+  }
+  // write the state of the task to the ProvisionerDataset, retrying if any exception is caught
+  protected void persistTaskInfo(ProvisioningTaskInfo taskInfo,
+                                 RetryStrategy retryStrategy) throws TransactionFailureException, InterruptedException {
+    try {
+      // Stop retrying if we are interrupted. Otherwise, retry on every exception, up to the retry limit
+      Retries.callWithInterruptibleRetries(() -> {
+        transactional.execute(dsContext -> {
+          ProvisionerDataset dataset = ProvisionerDataset.get(dsContext, datasetFramework);
+          dataset.putTaskInfo(taskInfo);
+        });
+        return null;
+      }, retryStrategy, t -> true);
+    } catch (TransactionFailureException e) {
+      // this is thrown if we ran out of retries
+      handleStateSaveFailure(taskInfo, e);
+      throw e;
+    }
+  }
+
+  protected abstract Map<ProvisioningOp.Status, ProvisioningSubtask> getSubtasks();
```

Comment:
Need javadoc for abstract method to document the contract for the children to implement.

Reply:
done